valhalla/thor/dijkstras.h

```diff
@@ -0,0 +1,273 @@
+
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <valhalla/baldr/double_bucket_queue.h>
+#include <valhalla/baldr/graphid.h>
+#include <valhalla/baldr/graphreader.h>
+#include <valhalla/baldr/location.h>
+#include <valhalla/proto/tripcommon.pb.h>
+#include <valhalla/sif/dynamiccost.h>
+#include <valhalla/sif/edgelabel.h>
+#include <valhalla/thor/edgestatus.h>
+
+namespace valhalla {
+namespace thor {
+
+enum class InfoEdgeType {
+  origin,
+  destination,
+  regular,
+};
+enum class InfoRoutingType {
+  forward,
+  bidirectional,
+  multi_modal,
+};
+
+struct ExpandingNodeMiscInfo {
+  InfoEdgeType edge_type;
+  InfoRoutingType routing_type;
+  bool from_transition;
+};
+
+enum class RouteCallbackRecommendedAction {
+  no_action,
+  skip_expansion,
+  stop_expansion,
+};
+
+/**
+ * Algorithms to do shortest first graph traversal
+ */
+class Dijkstras {
+public:
+  /**
+   * Constructor.
+   */
+  Dijkstras();
+
+  /**
+   * Destructor
+   */
+  virtual ~Dijkstras();
+
+  /**
+   * Clear the temporary memory (adjacency list, edgestatus, edgelabels)
+   */
+  void Clear();
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void Compute(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locs,
+               baldr::GraphReader& graphreader,
+               const std::shared_ptr<sif::DynamicCost>* mode_costing,
+               const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal to a list of destination locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeReverse(google::protobuf::RepeatedPtrField<valhalla::Location>& dest_locations,
+                      baldr::GraphReader& graphreader,
+                      const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                      const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations using multimodal
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeMultiModal(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locations,
+                         baldr::GraphReader& graphreader,
+                         const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                         const sif::TravelMode mode);
+
+protected:
+  // Virtual function called when expanding a node
+  //
+  // Children can implement this to customize behaviour
```

Comment:
Even children can implement this!?! :D Sorry, could not resist! :) No need to change, but if you do feel like, maybe use "derived" class instead.

Reply:
Hmm I thought parent-child terminology was established. Maybe I'm mixing it up with some other language.

Reply:
> Hmm I thought parent-child terminology was established. Maybe I'm mixing it up with some other language.

Don't worry, we don't need to change it. I've seen different terminology used at various places (child class, derived class, or subclass), so I guess this is one of those things. I was just making a cheeky comment! :)